src/Unleash/Scheduling/SystemTimerScheduledTaskManager.cs

```diff
@@ -80,34 +83,47 @@ async void Callback(object state)
             timers.Add(name, timer);
 
             // Now it's ok to start the timer.
-            timer.SafeTimerChange(dueTime, period, ref disposeEnded);
+            timer.SafeTimerChange(dueTime, period, ref _disposed);
         }
 
-        private bool disposeEnded;
+        private volatile bool _shuttingDown;
+        private bool _disposed;
+
         public void Dispose()
         {
-            if (disposeEnded)
-                return;
+            if (_disposed) return;
+            _shuttingDown = true;
 
             var timeout = TimeSpan.FromSeconds(1);
 
-            using (var waitHandle = new ManualResetEvent(false))
+            var keys = new List<string>(timers.Keys.AsEnumerable());

```

Comment:
Keys is a terrible name. I actually don't know what this is looking at it. Timer ids? Can we call it what it is?